### PR TITLE
Backend/model integration

### PR DESCRIPTION
**Pull request by conorato**
## Summary

This PR includes:
- changes to use the downloaded model from our storage in order to make predictions
- code to download the hidden markov model (hmm) parameters from our storage
- modifications as to how we check if the model is up to date. We now also check if the latest modification was made after the time the model or matrix was modified in the storage. What lead to this modification is that the files containing the parameters for the HMM model will most likely have the same size. Indeed, even if we retrain and change the parameters, the matrices or vector will have the same dimensions (either N_STAGE x N_STAGE, or N_STAGE elements) and the float size of each parameter will stay the same.
- Fixed a bug in the preprocessing steps, which led to all `W` predictions. Our training dataset is composed of raw EEG stored in EDF files. This file format stores, by default, amplitudes in Volts, and this scale was not changed afterwards. Thus, the prediction pipeline expects data to be in V, not in uV (which was the case in the backend before).

## To verify

We should check if the generated server can be executed with and without the `onnxruntime`'s dependancy, OpenMP.

### CODE REVIEW

**Comment by conorato** [comment]
[[1.1.6.1] Classification sur le serveur](https://trello.com/c/lrG12orI/62-1161-classification-sur-le-serveur)

File: backend/classification/load_model.py [hunk 1]
```diff
@@ -1,37 +1,91 @@
...
+
+
+def load_hmm():
+    hmm_matrices = dict()
+
+    if not path.exists(SCRIPT_PATH / HMM_FOLDER):
+        makedirs(SCRIPT_PATH / HMM_FOLDER)
+
+    for hmm_object_name, hmm_file in HMM_FILENAMES:
+        model_path = SCRIPT_PATH / HMM_FOLDER / hmm_file
```
Comment (abelfodil) [inline backend/classification/load_model.py:84]: nice syntactic sugar

File: backend/classification/load_model.py [hunk 2]
```diff
@@ -1,37 +1,88 @@
...
+def _get_latest_object_information(filename):
+    raw_result = get(BUCKET_URL, verify=certifi.where()).text
     # https://stackoverflow.com/a/15641319
     raw_result = re.sub(' xmlns="[^"]+"', '', raw_result)
     result_root_node = ET.fromstring(raw_result)
     objects_nodes = result_root_node.findall('Contents')
     object_node = [object_node for object_node in objects_nodes if object_node.find("Key").text == filename][0]
     object_size = int(object_node.find("Size").text)
-    return object_size
+    object_latest_update = datetime.strptime(object_node.find("LastModified").text, "%Y-%m-%dT%H:%M:%S.%f%z")
```
Comment (abelfodil) [inline backend/classification/load_model.py:49]: nice

Comment (abelfodil) [reply backend/classification/load_model.py:49]: Ce serait une métrique suffisante en fait non?

File: backend/app.py [hunk 3]
```diff
@@ -57,7 +62,7 @@ def analyze_sleep():
 
     try:
         raw_array = get_raw_array(file)
-        predict(raw_array, info={
+        predict(raw_array, model, info={
```
Comment (abelfodil) [inline backend/app.py:65]: Pourquoi ne pas faire une classe qui contient la méthode `predict` ?

File: backend/classification/load_model.py [hunk 4]
```diff
@@ -1,37 +1,88 @@
...
+MODEL_PATH = SCRIPT_PATH / MODEL_FILENAME
 MODEL_URL = f'{BUCKET_URL}/{MODEL_FILENAME}'
 
+HMM_FOLDER = 'hmm_model'
+HMM_FILENAMES = [
+    (HMM_EMISSION_MATRIX, 'hmm_emission_probabilites.npy'),
+    (HMM_START_PROBABILITIES, 'hmm_start_probabilities.npy'),
+    (HMM_TRANSITION_MATRIX, 'hmm_transition_probabilites.npy')
+]
+
```
Comment (abelfodil) [inline backend/classification/load_model.py:34]: nit: dict?

File: backend/classification/config/constants.py [hunk 5]
```diff
@@ -29,3 +29,9 @@ class Sex(Enum):
     [85, 125]
 ]
 ACCEPTED_AGE_RANGE = [AGE_FEATURE_BINS[0][0], AGE_FEATURE_BINS[-1][-1]]
+
+N_STAGES = 5
+
+HMM_EMISSION_MATRIX = 'emission'
+HMM_START_PROBABILITIES = 'start'
+HMM_TRANSITION_MATRIX = 'transition'
```
Comment (abelfodil) [inline backend/classification/config/constants.py:37]: nit: enum?

File: backend/classification/predict.py [hunk 6]
```diff
@@ -16,8 +20,18 @@ def predict(raw_eeg, info):
             the subject started the recording and went to bed
         - out_of_bed_seconds: timespan, in seconds, from which
             the subject started the recording and got out of bed
+    Returns: array of predicted sleep stages
     """
+    classifier, postprocessing_state = model['classifier'], model['postprocessing']
```
Comment (abelfodil) [inline backend/classification/predict.py:25]: this function knows too much imo

File: backend/classification/predict.py [hunk 7]
```diff
@@ -16,8 +20,18 @@ def predict(raw_eeg, info):
...
     validate(raw_eeg, info)
-    X_openbci = get_features(raw_eeg, info)
+    features = get_features(raw_eeg, info)
+    input_name = classifier.get_inputs()[0].name
+
+    predictions = classifier.run(None, {input_name: features})[0]
+    predictions = postprocess(predictions, postprocessing_state)
+
+    print(features[0], features.shape)
+    print(predictions)
```
Comment (abelfodil) [inline backend/classification/predict.py:35]: is this intentional?

File: backend/classification/load_model.py (hunk 4 above)
Comment (abelfodil) [reply backend/classification/load_model.py:34]: ou un enum direct en fait :O

File: backend/classification/postprocess.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    N_STAGES,
+)
+
+
+def postprocess(predictions, postprocessing_state):
+    hmm_model = MultinomialHMM(n_components=N_STAGES)
+
+    hmm_model.emissionprob_ = postprocessing_state[HMM_EMISSION_MATRIX]
+    hmm_model.startprob_ = postprocessing_state[HMM_START_PROBABILITIES]
+    hmm_model.transmat_ = postprocessing_state[HMM_TRANSITION_MATRIX]
```
Comment (abelfodil) [inline backend/classification/postprocess.py:16]: same, it knows too much

**Review by abelfodil — CHANGES_REQUESTED** [review]
Quelques commentaires par rapport à l'achitecture du modèle

File: backend/classification/predict.py (hunk 7 above)
Comment (conorato) [reply backend/classification/predict.py:35]: yes, as we still return the mock (will change to real data when working on calculating the metrics in another PR) and I wanted a way to see the results meanwhile.

File: backend/classification/load_model.py (hunk 4 above)
Comment (conorato) [reply backend/classification/load_model.py]: ahahah je ne penses jamais à utiliser des enum

**Review by abelfodil — APPROVED** [review]

**Pull request merged**